beacon-chain/blockchain/core.go

```diff
@@ -301,37 +326,85 @@ func (b *BeaconChain) computeNewActiveState(seed common.Hash) (*types.ActiveStat
 
 	// TODO: Verify randao reveal from validator's hash pre image.
 
-	return types.NewActiveState(&pb.ActiveState{
-		PendingAttestations: []*pb.AttestationRecord{},
-		RecentBlockHashes:   [][]byte{},
-	}), nil
+	return newActiveState, nil
+}
+
+// computeNewCrystallizedState for every newly processed beacon block at a cycle transition.
+func (b *BeaconChain) computeNewCrystallizedState(active *types.ActiveState, block *types.Block) (*types.CrystallizedState, error) {
+	newCrystallized := b.CrystallizedState()
+	newCrystallized.SetStateRecalc(block.SlotNumber())
+	if err := b.calculateRewardsFFG(active, newCrystallized, block); err != nil {
+		return newCrystallized, fmt.Errorf("could not calculate ffg rewards/penalties: %v", err)
+	}
+	return newCrystallized, nil
+}
+
+// calculateRewardsFFG adjusts validators balances by applying rewards or penalties
+// based on FFG incentive structure.
+func (b *BeaconChain) calculateRewardsFFG(active *types.ActiveState, crystallized *types.CrystallizedState, block *types.Block) error {
```

Comment:
Why the FFG suffix? Will there be multiple ways to calculate rewards?

Reply:
Is there a better name for this? CalculateRewards implies that this would be a pure function to calculate the rewards.

Maybe this method is doing too many things. Can it be refactored so that the calculate rewards method is a pure method? And maybe the Casper pure functions belong in their own package?